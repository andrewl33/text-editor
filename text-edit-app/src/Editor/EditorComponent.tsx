import * as React from "react";
// import * as Prism from 'prismjs';
import ContentEditable, { RefObject } from "react-sane-contenteditable";
import { Container, Grid } from "semantic-ui-react";
import Sidebar from "../generic/Sidebar/SidebarComponent";

import { EditorComponentProps } from "../types";
// import '../../node_modules/prismjs/themes/prism-okaidia.css';

interface EditorPrivateState {
  text: string;
}

// TODO: check https://github.com/ashleyw/react-sane-contenteditable/issues/31
// right now, the content editable does not listen to content correctly

// TODO: Update typings for react-sane-contenteditable
class EditorComponent extends React.Component<
  EditorComponentProps,
  EditorPrivateState
> {
  private delayBeforeUpdate: number;
  private myRef: RefObject<any>;
  constructor(props: EditorComponentProps) {
    super(props);

    this.myRef = React.createRef();
  }

  public render() {
    const {
      name,
      tags,
      createDate,
      isLocked,
      users,
      onRemoveTag,
      onAddTag,
      onAddUser,
      onRemoveUser,
      onNameChange,
      codeText
    } = this.props;

    return (
      <Grid style={{ marginTop: 0, height: "100%" }}>
        <Grid.Column width={13}>
          <div
            className="editor language-javascript"
            ref={this.myRef}
            style={{ height: "100%" }}
          >
            <ContentEditable
              className="code-line editor"
              content={codeText}
              editable={true}
              multiLine={true}
              onKeyDown={this.handleChange}
              style={{ padding: 0, height: "100%" }}
            />
          </div>
        </Grid.Column>
        <Grid.Column
          width={3}
          textAlign="center"
          style={{
            paddingRight: "2em",
            backgroundColor: "#3180d4",
            height: "100%"
          }}
        >
          <Container>
            <Sidebar
              name={name}
              pageType="file"
              tagList={tags}
              allTagsList={["woww", "samesame"]} // TODO:
              createDate={createDate}
              isPrivate={isLocked}
              users={users}
              onRemoveTag={onRemoveTag}
              onAddTag={onAddTag}
              onNameChange={onNameChange}
              onAddUser={onAddUser}
              onRemoveUser={onRemoveUser}
            />
          </Container>
        </Grid.Column>
      </Grid>
    );
  }

  private handleChange = (e: React.SyntheticEvent, value: string) => {
    this.setState({ text: value });
    this.props.onLocalUpdate(value);
    clearTimeout(this.delayBeforeUpdate);
    this.delay();
  };

  private delay = () => {
    // uses browser's setTimeout instead of node
    this.delayBeforeUpdate = window.setTimeout(() => {
      this.props.onBatchUpdate();
    }, 2500);
  };
}

export default EditorComponent;
