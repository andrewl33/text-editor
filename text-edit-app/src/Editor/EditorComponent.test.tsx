import { mount } from "enzyme";
import * as React from "react";
import ContentEditable from "react-sane-contenteditable";
import EditorComponent from "./EditorComponent";

const mockBatchUpdate = jest.fn();
const props = {
  codeText: "var wow",
  tags: [],
  isLoading: true,
  isNewPage: false,
  hasAuth: false,
  isLocked: false,
  isSaved: true,
  onBatchUpdate: mockBatchUpdate,
  openAlert: false,
  alertMessage: "",
  pathname: "test",
  authPrompt: false,
  filePrompt: false,
  createDate: "1-1-1990",
  users: ["jimmy", "joy"],
  name: "Example Lambdas",
  onCodeChange: () => {
    /**/
  },
  onAlert: () => {
    /**/
  },
  onLock: () => {
    /**/
  },
  onMount: () => {
    /**/
  },
  onNew: () => {
    /**/
  },
  onShare: () => {
    /**/
  },
  onAuthAccount: () => {
    /**/
  },
  onAuthFile: () => {
    /**/
  },
  onNameChange: () => {
    /**/
  }
};

const wrapper = mount(<EditorComponent {...props} />);

const contentEditable = wrapper.find(ContentEditable);

describe("EditorComponent", () => {
  test("it renders ContentEditable", () => {
    expect(contentEditable.length).toEqual(1);
  });

  test("it does not call mockFunction before 2.5 seconds", () => {
    contentEditable.simulate("change");
    expect(mockBatchUpdate).not.toBeCalled();
  });

  test("it does call mockFunction after 2.5 seconds", () => {
    contentEditable.simulate("change");
    setTimeout(() => {
      expect(mockBatchUpdate.mock.calls.length).toEqual(2);
    }, 2600);
  });

  test("delay resets the timer whenever another change is called", () => {
    for (let i = 0; i < 10; i++) {
      contentEditable.simulate("change");
    }

    expect(mockBatchUpdate.mock.calls.length).toBeLessThan(4);
  });
});
