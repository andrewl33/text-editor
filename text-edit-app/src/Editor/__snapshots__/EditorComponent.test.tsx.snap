// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditorComponent renders correctly 1`] = `
<div
  className="ui grid"
  style={
    Object {
      "height": "100%",
      "marginTop": 0,
    }
  }
>
  <div
    className="thirteen wide column"
    style={
      Object {
        "padding": 0,
      }
    }
  >
    <div
      style={
        Object {
          "height": "100%",
          "width": "100%",
        }
      }
    >
      <div
        id="ace-editor"
        style={
          Object {
            "height": "100%",
            "position": "absolute",
            "width": "100%",
          }
        }
      />
    </div>
  </div>
  <div
    className="center aligned three wide column"
    style={
      Object {
        "backgroundColor": "#3180d4",
        "height": "100%",
        "paddingRight": "2em",
      }
    }
  >
    <div
      className="ui container"
    >
      <div
        className="ui list"
        role="list"
        style={
          Object {
            "height": "100%",
          }
        }
      >
        <div
          className="item"
          onClick={[Function]}
          role="listitem"
        >
          <div
            className="header"
          >
            Name
          </div>
          <span
            className="name"
          >
            file name
            <i
              aria-hidden="true"
              className="edit link icon"
              onClick={[Function]}
            />
          </span>
        </div>
        <div
          className="item"
          onClick={[Function]}
          role="listitem"
        >
          <div
            className="header"
          >
            Date Created
          </div>
          1-1-1
        </div>
        <div
          className="item"
          onClick={[Function]}
          role="listitem"
        >
          <div
            className="header"
          >
            Private
          </div>
          No
        </div>
        <div
          className="item users"
          onClick={[Function]}
          role="listitem"
        >
          <div
            className="header"
          >
            Users
          </div>
          <div
            className="ui label"
            onClick={[Function]}
          >
            user1
          </div>
          <div
            className="ui label"
            onClick={[Function]}
          >
            user2
          </div>
          <span>
            <i
              aria-hidden="true"
              className="add link icon"
              onClick={[Function]}
            />
          </span>
        </div>
        <div
          className="item tags"
          onClick={[Function]}
          role="listitem"
        >
          <div
            className="header"
          >
            Tags
          </div>
          <span>
            <div
              className="ui label"
              link="true"
              onClick={[Function]}
            >
              tag1
              <i
                aria-hidden="true"
                className="close icon"
              />
            </div>
            <div
              className="ui label"
              link="true"
              onClick={[Function]}
            >
              tag2
              <i
                aria-hidden="true"
                className="close icon"
              />
            </div>
            <div
              className="ui label"
              link="true"
              onClick={[Function]}
            >
              tag3
              <i
                aria-hidden="true"
                className="close icon"
              />
            </div>
            <i
              aria-hidden="true"
              className="add link icon"
              onClick={[Function]}
            />
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;
